sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -1381,93 +1381,116 @@ ESIMD_NODEBUG
 //
 // inv, log2, exp2, sqrt, rsqrt, sin, cos
 //
-// share the same requirements.
-//
-// template <int SZ>
-// simd<float, SZ>
-// ESIMD_INLINE inv(simd<float, SZ> src0, int flag = saturation_off) {
-//   simd<float, SZ> Result = __esimd_inv(src0);
-//   if (flag != saturation_on)
-//     return Result;
-//   return __esimd_sat<float>(Result);
-// }
-//
-// template <int N1, int N2>
-// __ESIMD_API
-// simd<float, N1 * N2>
-// inv(matrix<float, N1, N2> src0, int flag = saturation_off) {
-//   simd<float, N1 * N2> Src0 = src0;
-//   return esimd::inv(Src0, flag);
-// }
-//
-// ESIMD_INLINE float inv(float src0, int flag = saturation_off) {
-//   simd<float, 1> Src0 = src0;
-//   simd<float, 1> Result = esimd::inv(Src0, flag);
-//   return Result[0];
-// }
-//
-// we also make the scalar version template-based by adding
-// a "typename T". Since the type can only be float, we hack it
-// by defining T=void without instantiating it to be float.
-
-#define ESIMD_INTRINSIC_DEF(type, name, iname)                                 \
-  template <int SZ>                                                            \
-  __ESIMD_API simd<type, SZ> name(simd<type, SZ> src0,                         \
-                                  int flag = saturation_off) {                 \
-    simd<type, SZ> Result = __esimd_##iname<SZ>(src0.data());                  \
+
+#define ESIMD_UNARY_INTRINSIC_DEF(COND, name, iname)                           \
+  /* Faster vector implementation w/o dynamic branch when saturation       */  \
+  /* parameter is known at compile-time.                                   */  \
+  template <class T, int N, int F = saturation_off,                            \
+            class = std::enable_if_t<COND>>                                    \
+  __ESIMD_API simd<T, N> name(simd<T, N> src) {                                \
+    __SEIEED::vector_type_t<__SEIEED::__raw_t<T>, N> res =                     \
+        __esimd_##iname<T, N>(src.data());                                     \
+    if constexpr (F != saturation_on)                                          \
+      return res;                                                              \
+    else                                                                       \
+      return esimd::saturate<T>(res);                                          \
+  }                                                                            \
+                                                                               \
+  /* Slower vector implementation with dynamic branch on saturation            \
+   * parameter.*/                                                              \
+  template <class T, int N, class = std::enable_if_t<COND>>                    \
+  __ESIMD_API simd<T, N> name(simd<T, N> src, int flag) {                      \
```

Comment:
Shouldn't 'flag' have the default value?
```suggestion
  __ESIMD_API simd<T, N> name(simd<T, N> src, int flag = saturation_off) {                      \
```

Reply:
no, this is intentional. When user calls 1-argument version, this will resolve to the efficient version above w/o the dynamic saturation flag check